exercises/react/example.cr

```diff
@@ -0,0 +1,72 @@
+class Cell
+  @dependents = [] of ComputeCell
+
+  def initialize(initial : Int32)
+    @value = initial
+  end
+
+  getter :value
+
+  protected getter :dependents
+end
+
+class InputCell < Cell
+  def value=(new_value : Int32)
+    @value = new_value
+    @dependents.each(&.propagate)
+    @dependents.each(&.fire_callbacks)
+  end
+end
+
+class ComputeCell < Cell
+  alias CallbackID = Int32
+
+  @callbacks_given = 0
+  @callbacks = {} of CallbackID => Int32 ->
+  @last_callback_value : Int32
+
+  def initialize(new_value : -> Int32, *inputs : Cell)
+    super(new_value.call)
+    @new_value = new_value
+    @last_callback_value = @value
+    inputs.each { |i| i.dependents << self }
+  end
+
+  def add_callback(&block : Int32 ->) : CallbackID
+    @callbacks_given += 1
+    @callbacks[@callbacks_given] = block
+    @callbacks_given
+  end
+
+  def remove_callback(id : CallbackID)
+    @callbacks.delete(id)
+  end
+
+  def propagate
+    v = @new_value.call
+    return if v == @value
+    @value = v
+    @dependents.each(&.propagate)
+  end
+
+  def fire_callbacks
+    return if @value == @last_callback_value
+    @last_callback_value = @value
+    @callbacks.each_value { |c| c.call(@value) }
+    @dependents.each(&.fire_callbacks)
+  end
+end
+
+class Reactor
```

Comment:
Note that this class does nothing *in this example solution*.

If we anticipate that *all submitted solutions* will do the same, then `create_input` and `create_compute` can be class methods such that they would be called as `React.create_input(1)` rather than `Reactor.new.create_input(1)`. However, in experience with this problem some valid implementations track information about each cell in the reactor. So I propose to leave it like this.